storage/engine/rocksdb/db.cc

```diff
@@ -1418,15 +1425,35 @@ DBIterator* DBBatch::NewIter(bool prefix) {
   return iter;
 }
 
-DBIterator* DBSnapshot::NewIter(bool prefix) {
-  DBIterator* iter = new DBIterator;
+DBIterator* DBSnapshot::NewIter(DBSlice prefix) {
+  DBIterator* iter = new DBIterator(prefix);
   rocksdb::ReadOptions opts = read_opts;
-  opts.total_order_seek = !prefix;
+  opts.iterate_upper_bound = iter->upper_bound_ptr;
+  opts.total_order_seek = iter->upper_bound_ptr == NULL;
   iter->rep.reset(rep->NewIterator(opts));
   return iter;
 }
 
-DBIterator* DBNewIter(DBEngine* db, bool prefix) {
+DBIterator::DBIterator(DBSlice prefix)
+    : upper_bound_ptr(NULL) {
+  if (prefix.len > 0) {
+    // When we're performing a prefix scan, we want to limit the scan
+    // to the keys that have the matching prefix. Prefix in this case
+    // refers to an exact match on the non-timestamp portion of a
+    // key. We do this be constructing an encoded mvcc key which has a
+    // zero timestamp (hence the trailing 0) and is the "next" key
+    // (thus the additional 0). See EncodeKey and SplitKey for more
+    // details on the encoded key format.
+    upper_bound_str.reserve(prefix.len + 2);
+    upper_bound_str.append(prefix.data, prefix.len);
+    upper_bound_str.push_back(0);
+    upper_bound_str.push_back(0);
+    upper_bound = upper_bound_str;
```

Comment:
I don't, but it makes setting `iterate_upper_bound` in the options more convenient.